### PR TITLE
Export speakers as PDF solves #5137

### LINKED ISSUE #5137: Export Speakers as PDF
**Is your feature request related to a problem? Please describe.**
Add an endpoint to export speakers as PDF

**Describe the solution you'd like**
Create a template and follow the steps used by other export endpoints.

**Describe alternatives you've considered**
\-

### PR DESCRIPTION
**Pull request by starlord1311**
Fixes #5137 

#### Checklist

- [x] I have read the [Contribution & Best practices Guide](https://blog.fossasia.org/open-source-developer-guide-and-best-practices-at-fossasia) and my PR follows them.
- [x] My branch is up-to-date with the Upstream `development` branch.
- [x] The unit tests pass locally with my changes 
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)

- [x] All the functions created/modified in this PR contain relevant docstrings.

#### Changes proposed in this pull request:

- Changes made in the export.py so as to add pdf for speakers.
-Speakers.html template created
-

### CODE REVIEW

**Comment by mayank8318** [comment]
@starlord1311 No, you didn't get me I think. You need to create a fresh template for displaying speaker info. Here you have just copied the orders template. You need to print the speakers, not the orders

**Comment by starlord1311** [comment]
@mayank8318 are the changes made in exports.py correct?

**Comment by starlord1311** [comment]
@mayank8318 can you check now?

**Comment by srv-twry** [comment]
I can't see the celery task either. @starlord1311 You have only declared the endpoint but not the task. @mayank8318 will guide you more on this.

File: app/templates/pdf/speakers.html [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+                            Payment Mode    :<br>
+                            Discount Code   :<br>
+                        {% if speaker.status == 'completed' %}
+                            Speaker Completed :<br>
+                        {% else %}
+                            Speaker Placed    :<br>
+                        {% endif %}
+                    </td>
+                    <td><br>
+                            <strong>{{ speaker.get_invoice_number() }}</strong><br>
```
Comment (mayank8318) [inline app/templates/pdf/speakers.html]: Speaker doesn't have an invoice number :( Just check the speaker.py file in models. Look at exporting speakers as csv. You need to print all those fields.

Comment (starlord1311) [reply app/templates/pdf/speakers.html]: exporting speakers as csv is similiar to exporting orders as csv. where to get all the fields?

Comment (mayank8318) [reply app/templates/pdf/speakers.html]: Have a look at #5156

Comment (mayank8318) [reply app/templates/pdf/speakers.html]: All the models are located in `app/models`

Comment (starlord1311) [reply app/templates/pdf/speakers.html]: I didnt understand how to get all the fields and which all fields to put

**Comment by starlord1311** [comment]
@mayank8318 can you check now?

File: app/api/exports.py [hunk 2]
```diff
@@ -263,6 +283,10 @@ def export_sessions_pdf(event_identifier):
     else:
         event_id = event_identifier
 
+    from .helpers.tasks import export_speakers_pdf_task
```
Comment (mayank8318) [inline app/api/exports.py]: Remove this please and fix the hound issue :)

**Comment by starlord1311** [comment]
@mayank8318 i didnt get the postman and screenshot part. what did you mean by it?

**Comment by mayank8318** [comment]
screenshot from 2018-07-22 13-02-19
This is the postman screenshot. It is an app by which you can make requests to the server.
And also add a screenshot of the pdf file generated.
Ps - To make the task work you will need to switch on Celery
`INTEGRATE_SOCKETIO=false celery worker -A app.celery`

**Comment by starlord1311** [comment]
what am i supposed to do in the postman app?

**Comment by starlord1311** [comment]
@mayank8318 how to get the pdf file? and where?

**Comment by mayank8318** [comment]
@starlord1311 Just follow these steps:
1. Start Celery by running this `INTEGRATE_SOCKETIO=false celery worker -A app.celery` in a terminal in the project directory. 
1. Create an event if you haven't created one already
1. Use that event's ID(identifier) and call the endpoint you created just now - `/events/<string:event_identifier>/export/speakers/pdf`. You need to do this in postman.
1. You will receive a task url like just like the one in the previous screenshot.
1. Make a call to this task URL
1. You will receive the download link. Open it in chrome and download the file

File: app/api/helpers/tasks.py [hunk 3]
```diff
@@ -388,6 +388,23 @@ def export_speakers_csv_task(self, event_id):
     return result
 
 
+@celery.task(base=RequestContextTask, name='export.speakers.pdf', bind=True)
+def export_speakers_pdf_task(self, event_id):
+    speakers = db.session.query(Speakers).filter_by(event_id=event_id)
```
Comment (mayank8318) [inline app/api/helpers/tasks.py]: The model name is `Speaker` and not `Speakers`.

Comment (mayank8318) [reply app/api/helpers/tasks.py]: It should be `speakers = db.session.query(Speaker).filter_by(event_id=event_id)`

File: app/api/helpers/tasks.py [hunk 4]
```diff
@@ -361,7 +361,7 @@ def export_sessions_csv_task(self, event_id):
 
 @celery.task(base=RequestContextTask, name='export.speakers.csv', bind=True)
 def export_speakers_csv_task(self, event_id):
-    speakers = db.session.query(Speaker).filter_by(event_id=event_id)
+    speaker = db.session.query(Speaker).filter_by(event_id=event_id)
```
Comment (mayank8318) [inline app/api/helpers/tasks.py]: Why change this?

**Comment by starlord1311** [comment]
@mayank8318 how to create an event?

File: app/templates/pdf/speakers.html [hunk 5]
```diff
@@ -0,0 +1,113 @@
+<!DOCTYPE html>
+<html lang="en">
+<head>
+	<meta charset="UTF-8">
+	<title>{{ event.name }} - {{ ("Speakers List") }}</title>
```
Comment (mayank8318) [inline app/templates/pdf/speakers.html]: This should be speakers[0].event.name

File: app/templates/pdf/speakers.html [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+			margin-right: 50px;
+		}
+
+		tr.padded td{
+		padding-top : 40px;
+		}
+	</style>
+</head>
+<body id="body">
+<h3 style="text-align:center;">{{ event.name }} - {{ ("Speakers List") }}</h3>
```
Comment (mayank8318) [inline app/templates/pdf/speakers.html]: Same here

**Comment by starlord1311** [comment]
screenshot from 2018-07-26 16-47-05
this is what postman shows @mayank8318 @srv-twry what to do now?

**Comment by mayank8318** [comment]
@starlord1311 Use the `Authentication` endpoint to get the token and then in the speakers' endpoint add a header as follows:
screenshot from 2018-07-26 16-56-52

Follow this method wherever JWT header is required.

**Comment by starlord1311** [comment]
@srv-twry @mayank8318 can you tell how to get access code. I put the authentication endpoint but didnt get any access code

**Comment by starlord1311** [comment]
can someone help?

**Comment by srv-twry** [comment]
@starlord1311 What's the issue that you faced while getting the access code ? Screenshot/ error message please.

**Comment by starlord1311** [comment]
@srv-twry i mentioned the problem on gitter. Got the access code but nothing happens

**Comment by srv-twry** [comment]
> Got the access code but nothing happens

What do you mean by nothing happens ? Use that token in future requests.

**Comment by starlord1311** [comment]
Yeah I used it but it doesn't work. I had sent a ss on gitter. Can you check it there please?

**Comment by srv-twry** [comment]
@starlord1311 Please finalize this PR by tomorrow otherwise @mayank8318 will go ahead and do it himself. This is needed urgently.

**Comment by starlord1311** [comment]
@srv-twry i had shown you the problem i am getting. what to do now?

**Comment by starlord1311** [comment]
@srv-twry apart from that i dont think this pr has any problems as such. so i guess it can be merged

**Comment by srv-twry** [comment]
It's an untested PR. We can't guide you step by step in solving issues. I told you to create an issue with the logs in the channel but I can't see any issue created by you. We can't merge this PR at all.

**Comment by starlord1311** [comment]
@srv-twry i had created an issue but a similar issue already existed so mine was closed

**Comment by srv-twry** [comment]
Alright, try again and see if it works now. It should be fixed now.

**Comment by starlord1311** [comment]
@srv-twry it is still showing 500 INTERNAL ERROR

**Comment by srv-twry** [comment]
post the stacktrace here.

**Comment by starlord1311** [comment]
@srv-twry stacktrace of postman or the webpage?

**Comment by starlord1311** [comment]
screenshot from 2018-08-06 23-46-04
@srv-twry the webpage stacktrace

**Comment by srv-twry** [comment]
this is an image. Why can't you copy that text and paste it here ?

**Comment by starlord1311** [comment]
@srv-twry how to define the event variable?

**Comment by srv-twry** [comment]
Have you created an event before creating a speaker ? If yes, then add it's id in the environment variables. If you haven't created it, then create it.

PS:
Please learn some REST API development in Flask first. You aren't aware of basic things. We can't do programming on your behalf.

**Comment by starlord1311** [comment]
@srv-twry I guess I havent created an event till how. How to create one?

**Comment by srv-twry** [comment]
Have a look at the api-docs. look for events section where you'll get an example on how to create event.

**Comment by starlord1311** [comment]
I got the id of the event as 1.
P.S - method to create event is to send a post request right?

**Comment by mayank8318** [comment]
@starlord1311 This PR still requires a lot of changes and needs to be finalized immediately. Seeing that you are still struggling with your setup, I am closing this PR now. I will be creating a PR for this shortly, you can see the changes there and compare :)

**Pull request closed without merging**